README.md
# Example Shiny app

This is an example application for Shiny.
The main purpose of this example is to illustrate how to run Shiny apps from a remote source.
There are many ways to download and run it:

First check the dependencies (you only need to do this once):

```R
if(!require(ggplot2)) install.packages("ggplot2")
if(!require(shiny)) install.packages("shiny")
```

Then load the shiny app:

```R
library(shiny)

# Easiest way is to use runGitHub
runGitHub("mirtools", "mariavica")
```

Other ways to load the app:

```R
# Run a tar or zip file directly
runUrl("https://github.com/mariavica/mirtools/archive/master.tar.gz")
runUrl("https://github.com/marivica/mirtools/archive/master.zip")

# Using runApp(),  first clone the repository with git. If you have cloned it into
# ~/shiny_example, first go to that directory, then use runApp().
setwd("~/mirtools")
runApp()
```


To run a Shiny app from a subdirectory in the repo or zip file, you can use the `subdir` argument. This repository happens to contain another copy of the app in `additional/mysubdir1`.

```R
runGitHub("mirtools", "mariavica", subdir = "additional/mysubdir1")

runUrl("https://github.com/mariavica/mirtools/archive/master.tar.gz",
  subdir = "additional/mysubdir1")


NEWS
VERSION 1.0.0
------------------------

    o First stable version.


server.r
shinyServer(function(input, output) {
   library(ggplot2)
   #versions.mirnas <- read.table("https://raw.githubusercontent.com/mariavica/mirtools/master/data/miRBase_conversions.csv",header=TRUE,sep="\t")
   #totrow<-nrow(versions.mirnas)
   #totcol<-ncol(versions.mirnas)
   load("mirbase_conversions.RData")
   
   species_name<-c("Homo sapiens","Mus musculus")
   prefix<-c("hsa-","mmu-")
   
    maketable <- reactive( if (input$mirname!=""  | !is.null(input$csvfile) )  {
      
      perc <- function (x,target) {
        return(length(which((target %in% x) == TRUE)) / length(target) *100)
      }
      
      versions.mirnas<-versions.mirnas[,-1]
      
      if (!is.null(input$csvfile)) {
        
        inFile<-input$csvfile
        
        uploadedm<-read.csv(inFile$datapath)
        
        mymirnas<-unlist(strsplit(as.character(uploadedm[,1]),c("\\,|\\ |\\\n")))
      }
      

      if (input$mirname!="") {
        mymirnas<-unlist(strsplit(as.character(input$mirname),c("\\,|\\ |\\\n")))
      }
      
    
      
      if (input$capitalise) {
        mymirnas<-gsub("[mM][iI][rR]","miR",mymirnas)
        mymirnas<-gsub("[Ll][Ee][Tt]","let",mymirnas)
        mymirnas<-gsub("A","a",mymirnas)
        mymirnas<-gsub("B","b",mymirnas)
        mymirnas<-gsub("C","c",mymirnas)
        mymirnas<-gsub("D","d",mymirnas)
        mymirnas<-gsub("E","e",mymirnas)
        mymirnas<-gsub("F","f",mymirnas)
        mymirnas<-gsub("G","g",mymirnas)
        mymirnas<-gsub("H","h",mymirnas)
        mymirnas<-gsub("I","i",mymirnas)
        mymirnas<-gsub("J","j",mymirnas)
      }
      
      
      if (input$species!="select") {
        specie<-which(species_name %in% input$species)
        sel<-c(grep("^miR",mymirnas),grep("^let",mymirnas))
        mymirnas[sel]<-paste(prefix[specie],mymirnas[sel],sep="")
        mymirnas[sel]<-gsub(paste(prefix[specie],prefix[specie],sep=""),prefix[specie],mymirnas[sel])
      } 
      

      mymirnas<-mymirnas[which(mymirnas!="")]
      #print(mymirnas)
      
      a<-apply(versions.mirnas,2,perc,mymirnas)
      
      dat<-data.frame(x=names(a),y=a)
      
      dat$x<-factor(gsub("miRBase_","",dat$x))
      dat$x<-relevel(dat$x,"9.2")
      dat$x<-relevel(dat$x,"9.1")
      dat$x<-relevel(dat$x,"9.0")
      dat$x<-relevel(dat$x,"8.2")
      dat$x<-relevel(dat$x,"8.1")
      dat$x<-relevel(dat$x,"8.0")
      dat$x<-relevel(dat$x,"7.1")
      dat$x<-relevel(dat$x,"7.0")
      dat$x<-relevel(dat$x,"6.0")
      #print(dat)
      
      maxs<-which(dat$y==max(dat$y))
      proposedversion<-dat[maxs[length(maxs)] ,"x"]	
      
      
      if (input$mirfrom != "I don't know") {
        selectedversion<-input$mirfrom
      } else {
        selectedversion<-proposedversion
      }
      
      #print(paste("miRBase_",selectedversion,sep=""))
      
      mymirnas<-as.character(mymirnas)
      mytrans<-data.frame(mymirnas)
      
      for (i in 1:nrow(mytrans)) {
        if (mymirnas[i] %in% versions.mirnas[,paste("miRBase_",selectedversion,sep="")] ) {
          mytrans[i,2]<-as.character(versions.mirnas[ which(  (versions.mirnas[,paste("miRBase_",selectedversion,sep="")] %in%  mymirnas[i])  ), c(paste("miRBase_",as.character(input$mirto),sep="")) ])
          
        } else {
          
          if (mymirnas[i] %in% as.vector(as.matrix(versions.mirnas))) {
            
            if (input$forceTranslation) {
                  coincidences <- which(t(versions.mirnas)==mymirnas[i])
                  #print(coincidences)
                  rowmir <- ceiling(coincidences[length(coincidences)]/(totcol-1))
                  #print(rowmir)
                  
                  mytrans[i,2]<-as.character(versions.mirnas[rowmir, c(paste("miRBase_",as.character(input$mirto),sep="")) ])
            }
            
            else {
            
            coincidences <- which(versions.mirnas==mymirnas[i])
            version <- colnames(versions.mirnas)[ceiling(coincidences[length(coincidences)]/totrow)]
            
             mytrans[i,2]<-as.character(paste("Not found in miRBase_",selectedversion," (found in ", version,")", sep=""))
             
            }
             
          } else {
            mytrans[i,2]<-paste("Unknown miRNA")
          }
          
        }
      }
      
      colnames(mytrans)<-c((paste("miRBase_",as.character(selectedversion),sep="")),c(paste("miRBase_",as.character(input$mirto),sep="")))
      
      
      return(list(mytrans,proposedversion,dat))
    })
    
    output$text1 <- renderText({
      paste("Most of your miRNAs are from version:",maketable()[[2]],"\n")
    })  
    
    output$percent <- renderPlot({
      datp<-maketable()[[3]]
      qplot(x=datp$x, y=datp$y, fill=datp$x) + geom_bar(stat="identity") +
        guides(fill=FALSE) + xlab("miRBase version") + ylab("Coincidence (%)")
    })

    output$translated<-renderTable(maketable()[[1]])
    
    output$downloadTranslated <- downloadHandler( filename="translated.csv", content=function (file){ write.table(maketable()[[1]], file, row.names=FALSE, sep="\t", quote=FALSE) })
    

    
})


ui.r
shinyUI(fluidPage(
  titlePanel("MiRNA version identifier and translator"),
  
  sidebarLayout(
    sidebarPanel(
      
      helpText("Upload your file (\"*txt\" or \"*csv\" plain file):"),
    
      fileInput(inputId="csvfile", label=NULL , multiple = FALSE, accept = c(
        "text/csv", "text/comma-separated-values,text/plain",".csv",".txt"), width = NULL,
        buttonLabel = "Browse...", placeholder = "No file selected"),
      
      helpText("or paste here your miRNAs to start:"),
      textAreaInput(inputId = "mirname", label=NULL,cols=20,rows=17),
  #    helpText("or update your file"),
  #    fileInput(inputId = "mirfile", label=NULL, multiple = FALSE, accept = NULL, width = NULL, 
  #              buttonLabel = "Browse...", placeholder = "No file selected"),

      helpText("Your miRBase version"),
            
      selectInput(inputId = "mirfrom", label=NULL, choices=c("I don't know", "6.0", "7.0", "7.1", "8.0", "8.1", "8.2", "9.0", "9.1", "9.2", "10.0", "10.1", "11.0", "12.0", "13.0", "14", "15", "16", "17", "18", "19", "20", "21" ),
                  selected = "I don't know" , multiple = FALSE, selectize = TRUE, width = NULL, size = NULL),
      
      helpText("Your desired miRBase version"),
      selectInput(inputId = "mirto", label=NULL, choices=c("6.0", "7.0", "7.1", "8.0", "8.1", "8.2", "9.0", "9.1", "9.2", "10.0", "10.1", "11.0", "12.0", "13.0", "14", "15", "16", "17", "18", "19", "20", "21" ),
                  selected = "21", multiple = FALSE, selectize = TRUE, width = NULL, size = NULL),

   helpText(strong("Proofreading:")),

   checkboxInput(inputId="forceTranslation",label="Force translation (miRNAs from other versions will be automatically translated to desired version)", value=FALSE),


  helpText("MiRNA species"),
  
  selectInput(inputId = "species", label=NULL, choices=c("select","Homo sapiens","Mus musculus"),
              selected = "Specified in name", multiple = FALSE, selectize = TRUE, width = NULL, size = NULL),
  

  checkboxInput(inputId="capitalise",label="Correct capital letters (valid only for \"miR\" and \"let\")", value=FALSE)#,
  
#  helpText("Substitute"),
#  textInput("subst.orig",label=NULL,value=NULL),
#  textInput("subst.fin",label=NULL,value="")

    ),    
    
    mainPanel(
      #textOutput("text2", container=span),
      helpText("Coincidences across miRNA versions"),
      plotOutput("percent"),
      textOutput("text1", container=span),
      helpText("Your translated miRNAs are:"),
      downloadButton("downloadTranslated",label="Download table"),
      tableOutput("translated")
    )
  )  
))
